R/unnecessary_nesting_linter.R

```diff
@@ -0,0 +1,180 @@
+#' Block instances of excessive nesting
+#'
+#' Excessive nesting harms readability. Use helper functions or early returns
+#'   to reduce nesting wherever possible.
+#'
+#' @param allow_assignment Logical, default `TRUE`, in which case
+#'   braced expressions consisting only of a single assignment are skipped.
+#'   if `FALSE`, all braced expressions with only one child expression are linted.
+#'   The `TRUE` case facilitates interaction with [implicit_assignment_linter()]
+#'   for certain cases where an implicit assignment is necessary, so a braced
+#'   assignment is used to further distinguish the assignment. See examples.
+#'
+#' @examples
+#' # will produce lints
+#' code <- "if (A) {\n  stop('A is bad!')\n} else {\n  do_good()\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' code <- "tryCatch(\n  {\n    foo()\n  },\n  error = identity\n)"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' code <- "expect_warning(\n  {\n    x <- foo()\n  },\n  'warned'\n)"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter(allow_assignment = FALSE)
+#' )
+#'
+#' # okay
+#' code <- "if (A) {\n  stop('A is bad because a.')\n} else {\n  stop('!A is bad too.')\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' code <- "capture.output({\n  foo()\n})"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' code <- "expect_warning(\n  {\n    x <- foo()\n  },\n  'warned'\n)"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' @evalRd rd_tags("unnecessary_nesting_linter")
+#' @seealso
+#'  - [cyclocomp_linter()] for another linter that penalizes overly complexcode.
+#'  - [linters] for a complete list of linters available in lintr.
+#' @export
+unnecessary_nesting_linter <- function(allow_assignment = TRUE) {
+  exit_calls <- c("stop", "return", "abort", "quit", "q")
+  exit_call_expr <- glue("
+    expr[SYMBOL_FUNCTION_CALL[{xp_text_in_table(exit_calls)}]]
+  ")
+  # block IF here for cases where a nested if/else is entirely within
+  #   one of the branches.
+  # TODO(michaelchirico): we could try and make the parallel exits requirement
```

Comment:
Track this in an issue instead?

Reply:
I'm actually not sure anymore what I meant here, exactly. Just deleting for now, it'll come up again if really important.